### PR TITLE
Allow using maps instead of structs

### PR DESCRIPTION
**Pull request by pcarranzav**
Sometimes we want to deal with objects with dynamic fields, especially in resin-supervisor where we only want to set certain fields, and sometimes we want to set fields as null.

Adding omitempty to some fields in the resin models would allow us to not send certain fields, but would not allow to send them as null values.
Conversely, changing fields in the resin models to pointers would allow us to send fields as null, but would not allow us to not send certain fields at all.

By allowing users of this library to pass the request objects as a `map[string]interface{}`, we give the flexibility to either not include a field (by not adding the key to the map) or to send it as null (by setting the value of that key to `nil`.

We add this option to all the REST methods, which gives users the possibility to use this library only with maps, and without knowledge of all the fields in the resin models.

### CODE REVIEW

File: resource.go [hunk 1]
```diff
@@ -25,24 +25,46 @@ func resourceNameFromStruct(v interface{}) string {
 	return strings.ToLower(structs.Name(v))
 }
 
+func resourceNameFromMap(v interface{}) (string, error) {
+	if m, ok := v.(map[string]interface{}); !ok {
+		return "", errors.New("Invalid map type")
+	} else if val, ok := m["pinejs"].(string); ok && val != "" {
+		return val, nil
```
Comment (lorenzo-stoakes) [inline resource.go]: It might be a little easier to parse this logic if you swapped this to `!ok || val == ""` and switched the returns, as ATM the logic is 'if error, report, else if no other error succeed else report other error', whereas 'if error report, else if other error report else succeed' feels a bit clearer.

File: resource.go [hunk 2]
```diff
@@ -25,24 +25,46 @@ func resourceNameFromStruct(v interface{}) string {
...
-
-		// Deref the pointer and recurse on that value until we get to a struct.
-		el := ptr.Elem().Interface()
-		return resourceName(el)
+	case reflect.Map:
+		return resourceNameFromMap(v)
+	case reflect.Ptr:
+		return resourceName(reflect.Indirect(reflect.ValueOf(v)).Interface())
+	case reflect.Slice:
+		if ty.Elem().Kind() == reflect.Struct {
```
Comment (lorenzo-stoakes) [inline resource.go]: It'd be nice to separate this logic out into a function as with the other cases + make this a little more readable.

File: resource.go [hunk 3]
```diff
@@ -25,24 +25,46 @@ func resourceNameFromStruct(v interface{}) string {
...
+			if theSlice.Len() == 0 {
+				return "", errors.New("Slice of maps must have non-zero length")
+			} else {
+				return resourceName(theSlice.Index(0).Interface())
+			}
+		}
 	}
 
-	return "", fmt.Errorf("tried to retrieve resource name from non-struct %s",
+	return "", fmt.Errorf("tried to retrieve resource name from non-struct and non-map %s",
```
Comment (lorenzo-stoakes) [inline resource.go]: I'm loving how you're updating everything to refer to the correct thing @pcarranzav very nice work! :)

File: resource.go [hunk 4]
```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
 	return
 }
 
+func getIdFromMap(m map[string]interface{}) (ret int, err error) {
```
Comment (lorenzo-stoakes) [inline resource.go]: This function never returns `err` in any case, it's nice that you can at line 105 below just `return`, but I think it'd be clearer to remove the `err` from here and change that line.

File: resource.go [hunk 5]
```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
...
+	if val, ok := m["id"].(int); !ok {
```
Comment (lorenzo-stoakes) [inline resource.go]: I might be losing track of my own original code here a bit :P but, is this always lower case?

File: resource.go [hunk 6]
```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
...
+	} else {
+		ret = val
+	}
+	return
+}
+
 // Retrieve Id field from interface.
 func resourceId(v interface{}) (ret int, err error) {
 	var f *structs.Field
+	invalidTypeErr := errors.New("Not a struct, map, or pointer to either")
```
Comment (lorenzo-stoakes) [inline resource.go]: Why declare this as a separate variable? I think it'd be fine set directly on L114 as `err := errors.New(...)`.

File: resource.go [hunk 7]
```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
...
+	ty := reflect.TypeOf(v)
 
-	if f, err = getResourceField(v); err == nil {
-		ret = f.Value().(int)
+	switch ty.Kind() {
```
Comment (lorenzo-stoakes) [inline resource.go]: Why keep `ty` around? also, you can do `switch ty := reflect.TypeOf(v); ty.Kind()` regardless, but I think `switch reflect.TypeOf(v) {` would be nice.

In case you weren't aware, you can just avoid all of this reflection code and do a straight up [type switch](https://golang.org/doc/effective_go.html#type_switch) like:

```go
switch v := v.(type) {
    case int:
        ... deal with int ...
    case string:
        ... deal with string ...   
    ... etc. ...
}
```

But here since you're trying to deal with arbitrary structs + maps I think that wouldn't be appropriate. Mentioning it in case you hadn't seen it before :)

File: util.go [hunk 8]
```diff
@@ -75,13 +75,23 @@ func isPointerToStruct(v interface{}) (bool, error) {
...
+}
+
+func isPointerToSliceStructsOrMaps(v interface{}) (bool, error) {
 	if el, err := ptrType(v); err != nil {
 		return false, err
 	} else if el.Kind() != reflect.Slice {
 		return false, errors.New("not a pointer to a slice")
-	} else if el.Elem().Kind() != reflect.Struct {
-		return false, errors.New("not a pointer to a slice of structs")
+	} else if el.Elem().Kind() != reflect.Struct && el.Elem().Kind() != reflect.Map {
```
Comment (lorenzo-stoakes) [inline util.go]: Since we check twice would be nicer to declare the element kind, e.g. `else if k := el.Elem().Kind(); k != reflect.Struct && k != reflect.Map {`

File: test/test.go [hunk 9]
```diff
@@ -2,36 +2,72 @@ package main
 
 import (
```
Comment (lorenzo-stoakes) [inline test/test.go]: Improved testing love it :)

File: resource.go [hunk 10]
```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
...
+	case reflect.Ptr:
+		return resourceId(reflect.Indirect(reflect.ValueOf(v)).Interface())
+	default:
+		err = invalidTypeErr
 	}
 
 	return
 }
+
+// Determine whether the struct's id will be omitted in json encoding.
```
Comment (lorenzo-stoakes) [inline resource.go]: Why was this moved from `utils.go`? Seems like that's the better place for this function

File: resource.go [hunk 11]
```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
...
+func isIdOmitted(v interface{}) (bool, error) {
+
+	if structs.IsStruct(v) {
```
Comment (lorenzo-stoakes) [inline resource.go]: I always think `if ... { <big chunk of code> } else { <other big chunk of code> }` is an indicator of the need for a refactor - it's harder to read and this seems to me to indicate that each block of code is doing something distinct which could be nicely self-documented with a separate function.

Perhaps moving to `isStructIdOmitted()` and `isMapIdOmitted()` would work better?

File: resource.go [hunk 12]
```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
...
+		} else {
+			// getResourceField() ensures this is an int.
+			id := f.Value().(int)
+			// Json tags are comma separated. 'omitempty' means id == 0 -> id field
+			// not included in generated json. Also note spacing, like "id,
+			// omitempty" is significant and prevents a tag from taking effect so no
+			// need for trimming.
+			//
+			// See http://golang.org/pkg/encoding/json/#Marshal
+			for _, field := range strings.Split(jsonTag, ",") {
```
Comment (lorenzo-stoakes) [inline resource.go]: My spider sense tells me maybe something's wrong here, which by all accounts could totally be down to abominations I've done in the original code :P it seems like we're trying to duplicate some go JSON marshalling logic here, and it seems quite inefficient. Also if this is just for the ID field which the comment suggests, surely we can just look at the field name?

I might be wrong, however :)

Comment (pcarranzav) [inline resource.go]: This is indeed part of your original code, haven't touched it, only moved it here as I perceived it was directly related to the "resource" - I can move it back though

**Comment by pcarranzav** [comment]
@lorenzo-stoakes I'll merge this as-is since your comments, albeit accurate, are mostly stylistic and I need to use this package rather urgently. I'll address your comments on a separate PR as soon as I have the time.
